lib/SILGen/SILGenExpr.cpp

```diff
@@ -3549,7 +3549,30 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
     [this, &operands, E](const KeyPathExpr::Component &component) {
       if (!component.getIndexExpr())
         return;
-      
+
+      if (auto *parenExpr = dyn_cast<ParenExpr>(component.getIndexExpr())) {
+        if (auto *defaultArg =
+                dyn_cast<DefaultArgumentExpr>(parenExpr->getSubExpr())) {
+          const ParamDecl *defaultParam = getParameterAt(
+              cast<ValueDecl>(defaultArg->getDefaultArgsOwner().getDecl()), 0);
+          parenExpr->setSubExpr(defaultParam->getDefaultValue());
+        }
+      }
+
+      if (auto *tupleExpr = dyn_cast<TupleExpr>(component.getIndexExpr())) {
+        size_t count = 0;
+        for (auto *element : tupleExpr->getElements()) {
```

Comment:
Actually here is one more refactoring suggestion:

```c++
for (unsigned i = 0, n = tupleExpr->getNumElements(); i != n; ++i) {
  const auto &element = tupleExpr->getElement(i);
  if (auto *defaultArg = dyn_cast<DefaultArgumentExpr>(element)) {
    // do the thing
  }
}
```

Reply:
Should we be modifying the AST in SILGen like this? Maybe we could do this without modifying the AST here? or perhaps do this in CSApply perhaps.

Reply:
I'm not expert in SILGen and was is going on there so @jckarter should take a look at this. Looking at `CSApply` `coerceCallArguments` should build a new index expression for subscript component which should include `DefaultArgumentExpr`.

Reply:
Yes. It looks like the following for the above example:
```
(paren_expr implicit type='(Int)' location=run.swift:19:10 range=[run.swift:19:10 - line:19:11]
  (default_argument_expr implicit type='Int' location=run.swift:19:10 range=[run.swift:19:10 - line:19:10] default_args_owner=run.(file).Foo.subscript(_:)@run.swift:16:3 param=0))
```

Reply:
I'll try to move the above default argument parsing into `coerceCallArguments`.

Reply:
When I moved it to `coerceCallArguments` I got errors in `visitIdentityExpr` (because `getType` has an assertion). Any ideas or, should I move it back to `SILGenExpr.cpp`?

Reply:
I think `coeceCallArguments` does the right thing for keypath component, just like for other expressions with default arguments. Maybe it makes sense to check how regular subscript calls with default arguments are lowered to see if it does make sense to do the transformation in SILGen?

Reply:
Regular subscripts call `buildSubscriptHelper` which calls `coerceCallArguments` and that handles the default args. Interestingly, we also call `coeceCallArguments` but the default parameters aren't lowered. I'll look into why this is.

Reply:
I've moved the default argument lowering into [`coerceCallArguments`](https://github.com/apple/swift/pull/27951/files#diff-cf676abf2776cbcdb21380661ec41563R5403-R5411). It seems to work but, I am curious as to why this wasn't done before and suspect it may be because there are some cases where lowering the default args in `coerceCallArguments` doesn't work (that is why I kept the creation of `DefaultArgumentExpr` as a fallback).